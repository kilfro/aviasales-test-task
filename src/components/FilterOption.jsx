import '../style/filter-option.css'

import Checkbox from './Checkbox'
import PropTypes from 'prop-types'
import React from 'react'

const FilterOption = ({ value, ...rest }) => {
    return (
        <label htmlFor={rest.id} className='filter-option'>
            <Checkbox {...rest} />{value}
        </label>
    )
}

FilterOption.propTypes = {
    checked: PropTypes.bool.isRequired,
    id: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    value: PropTypes.string.isRequired,
}

export default FilterOption