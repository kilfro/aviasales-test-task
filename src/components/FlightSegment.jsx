import { getDurationString, getFlightTimes, getStopsQuantityString } from '../functions'

import React from 'react'

const FlightSegment = ({ origin, destination, date, duration, stops }) => (
    <div className='flight-segment'>
        <div className='segment-part'>
            <div>{origin} - {destination}</div>
            <div>{getFlightTimes(new Date(date), duration)}</div>
        </div>
        <div className='segment-part'>
            <div>в пути</div>
            <div>{getDurationString(duration)}</div>
        </div>
        <div className='segment-part'>
            <div>{getStopsQuantityString(stops.length)}</div>
            <div>{stops.join(', ')}</div>
        </div>
    </div>
)

export default FlightSegment