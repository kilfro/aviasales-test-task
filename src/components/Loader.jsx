import '../style/loader.css'

import React from 'react'
import logo from '../logo.png'

const Loader = () => <img src={logo} className='loader' />

export default Loader