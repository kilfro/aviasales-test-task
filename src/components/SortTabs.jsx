import '../style/sort-tabs.css'

import React from 'react'
import { connect } from 'react-redux'
import { setOrderBy } from '../store/actions'

const SortTabs = ({ orderBy, setOrderBy }) => {
    const orderOptions = {
        price: 'Самый дешевый',
        duration: 'Самый быстрый',
    }

    const chosenStyle = {
        border: 'none',
        background: '#2196F3',
        color: '#FFFFFF',
    }

    const getChosenStyle = (option) => orderBy === option ? chosenStyle : {}

    return (
        <div className='tabs-container'>
            {Object.entries(orderOptions).map(([key, label]) => (
                <div
                    style={getChosenStyle(key)}
                    onClick={() => setOrderBy(key)}
                    key={key}
                >
                    {label}
                </div>
            ))}
        </div>
    )
}

const mapStateToProps = state => ({
    orderBy: state.orderBy
})

const mapDispatchToProps = {
    setOrderBy
}

export default connect(mapStateToProps, mapDispatchToProps)(SortTabs)