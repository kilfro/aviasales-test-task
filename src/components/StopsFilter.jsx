import '../style/stops-filter.css'

import FilterOption from './FilterOption'
import React from 'react'

const all = new Map()
all.set(0, 'Без пересадок')
all.set(1, '1 пересадка')
all.set(2, '2 пересадки')

const StopsFilter = ({ allOptions = all }) => {
    const [filterBy, setFilterBy] = React.useState([])
    const [chooseAll, setChooseAll] = React.useState(false)

    const isChecked = key => filterBy.includes(key)

    const changeHandler = key => {
        let newFilterBy
        if (isChecked(key)) {
            newFilterBy = filterBy.filter(f => f !== key)
            if (chooseAll) {
                setChooseAll(false)
            }
        } else {
            newFilterBy = [...filterBy, key]
            if (allOptions.size === newFilterBy.length) {
                setChooseAll(true)
            }
        }

        setFilterBy(newFilterBy)
    }

    const chooseAllHandler = () => {
        if (chooseAll) {
            setChooseAll(false)
            setFilterBy([])
        } else {
            setChooseAll(true)
            setFilterBy(Array.from(allOptions.keys()))
        }
    }

    return <div className='filter-container'>
        <div style={{margin: '0 20px 10px 20px'}}>количество пересадок</div>
        <FilterOption checked={chooseAll} onChange={chooseAllHandler} value='Все' />
        {Array.from(allOptions).map(([key, value]) => (
            <FilterOption id={key.toString()} checked={isChecked(key)} onChange={() => changeHandler(key)} value={value} key={key.toString()} />
        ))}
    </div>
}

export default StopsFilter