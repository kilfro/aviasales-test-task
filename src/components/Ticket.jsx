import '../style/ticket.css'

import FlightSegment from './FlightSegment'
import React from 'react'

const Ticket = ({ price, carrier, segments }) => {
    return (
        <div className='ticket-card'>
            <div className='ticket-header'>
                <div>{price.toLocaleString('ru-ru')} P</div>
                <img src={`http://pics.avs.io/99/36/${carrier}.png`} />
            </div>
            {segments.map(segment => <FlightSegment key={segment.date} {...segment} />)}
        </div>
    )
}

export default Ticket