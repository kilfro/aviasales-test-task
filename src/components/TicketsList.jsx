import '../style/tickets-list.css'

import React from 'react'
import SortTabs from './SortTabs'
import Ticket from './Ticket'
import { connect } from 'react-redux'

export const TicketsList = ({ tickets, stopsFilter, orderBy }) => {
    const filterBySteps = ticket => {
        if (stopsFilter.length === 0) return true
        return ticket.segments.reduce((sum, segment) => stopsFilter.includes(segment.stops.length) && sum, true)
    }

    const sumUpDuration = ticket => {
        return ticket.segments.reduce((sum, next) => sum + next.duration, 0)
    }

    const comparator = (one, two) => {
        if (orderBy === 'duration') {
            return sumUpDuration(one) - sumUpDuration(two)
        }
        return one[orderBy] - two[orderBy]
    }

    return (
        <div className='tickets-list'>
            <SortTabs />
            {tickets
                .filter(filterBySteps)
                .sort(comparator)
                .slice(0, 20)
                .map(ticket => <Ticket {...ticket} key={JSON.stringify(ticket)} />)}
        </div>
    )
}

const mapStateToProps = state => ({
    tickets: state.tickets,
    stopsFilter: state.stopsFilter,
    orderBy: state.orderBy,
})

export default connect(mapStateToProps)(TicketsList)