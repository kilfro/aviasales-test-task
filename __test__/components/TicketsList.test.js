import React from 'react'
import { TicketsList } from '../../src/components/TicketsList'

describe(TicketsList, () => {
  const tickets = [
    {
      price: 31101,
      carrier: 'SU',
      segments: [
        {
          origin: 'MOW',
          destination: 'HKT',
          date: '2021-04-03T02:33:00.000Z',
          stops: [],
          duration: 1729,
        },
        {
          origin: 'HKT',
          destination: 'MOW',
          date: '2021-04-23T01:08:00.000Z',
          stops: ['SHA', 'BKK'],
          duration: 1149,
        },
      ],
    },
    {
      price: 44751,
      carrier: 'EY',
      segments: [
        {
          origin: 'MOW',
          destination: 'HKT',
          date: '2021-04-03T01:37:00.000Z',
          stops: ['IST', 'HKG'],
          duration: 670,
        },
        {
          origin: 'HKT',
          destination: 'MOW',
          date: '2021-04-23T15:13:00.000Z',
          stops: ['SIN', 'DXB'],
          duration: 1289,
        },
      ],
    },
    {
      price: 30982,
      carrier: 'EK',
      segments: [
        {
          origin: 'MOW',
          destination: 'HKT',
          date: '2021-04-03T06:30:00.000Z',
          stops: ['AUH', 'SHA', 'BKK'],
          duration: 1902,
        },
        {
          origin: 'HKT',
          destination: 'MOW',
          date: '2021-04-23T05:32:00.000Z',
          stops: ['BKK', 'IST'],
          duration: 941,
        },
      ],
    },
    {
      price: 79466,
      carrier: 'SU',
      segments: [
        {
          origin: 'MOW',
          destination: 'HKT',
          date: '2021-04-03T04:20:00.000Z',
          stops: ['SIN', 'SHA'],
          duration: 661,
        },
        {
          origin: 'HKT',
          destination: 'MOW',
          date: '2021-04-23T05:53:00.000Z',
          stops: ['AUH'],
          duration: 1600,
        },
      ],
    },
    {
      price: 43073,
      carrier: 'TG',
      segments: [
        {
          origin: 'MOW',
          destination: 'HKT',
          date: '2021-04-03T08:09:00.000Z',
          stops: ['IST', 'BKK', 'DXB'],
          duration: 855,
        },
        {
          origin: 'HKT',
          destination: 'MOW',
          date: '2021-04-22T23:47:00.000Z',
          stops: [],
          duration: 1750,
        },
      ],
    },
  ]

  test('must be rendered correctly', () => {
    const component = shallow(
      <TicketsList tickets={tickets} orderBy='price' stopsFilter={[]} />
    )

    expect(component).toMatchSnapshot()
  })
})
