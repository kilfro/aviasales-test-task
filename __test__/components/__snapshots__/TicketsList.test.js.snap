// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketsList must be rendered correctly 1`] = `
<div
  className="tickets-list"
>
  <Connect(SortTabs) />
  <Ticket
    carrier="EK"
    key="{\\"price\\":30982,\\"carrier\\":\\"EK\\",\\"segments\\":[{\\"origin\\":\\"MOW\\",\\"destination\\":\\"HKT\\",\\"date\\":\\"2021-04-03T06:30:00.000Z\\",\\"stops\\":[\\"AUH\\",\\"SHA\\",\\"BKK\\"],\\"duration\\":1902},{\\"origin\\":\\"HKT\\",\\"destination\\":\\"MOW\\",\\"date\\":\\"2021-04-23T05:32:00.000Z\\",\\"stops\\":[\\"BKK\\",\\"IST\\"],\\"duration\\":941}]}"
    price={30982}
    segments={
      Array [
        Object {
          "date": "2021-04-03T06:30:00.000Z",
          "destination": "HKT",
          "duration": 1902,
          "origin": "MOW",
          "stops": Array [
            "AUH",
            "SHA",
            "BKK",
          ],
        },
        Object {
          "date": "2021-04-23T05:32:00.000Z",
          "destination": "MOW",
          "duration": 941,
          "origin": "HKT",
          "stops": Array [
            "BKK",
            "IST",
          ],
        },
      ]
    }
  />
  <Ticket
    carrier="SU"
    key="{\\"price\\":31101,\\"carrier\\":\\"SU\\",\\"segments\\":[{\\"origin\\":\\"MOW\\",\\"destination\\":\\"HKT\\",\\"date\\":\\"2021-04-03T02:33:00.000Z\\",\\"stops\\":[],\\"duration\\":1729},{\\"origin\\":\\"HKT\\",\\"destination\\":\\"MOW\\",\\"date\\":\\"2021-04-23T01:08:00.000Z\\",\\"stops\\":[\\"SHA\\",\\"BKK\\"],\\"duration\\":1149}]}"
    price={31101}
    segments={
      Array [
        Object {
          "date": "2021-04-03T02:33:00.000Z",
          "destination": "HKT",
          "duration": 1729,
          "origin": "MOW",
          "stops": Array [],
        },
        Object {
          "date": "2021-04-23T01:08:00.000Z",
          "destination": "MOW",
          "duration": 1149,
          "origin": "HKT",
          "stops": Array [
            "SHA",
            "BKK",
          ],
        },
      ]
    }
  />
  <Ticket
    carrier="TG"
    key="{\\"price\\":43073,\\"carrier\\":\\"TG\\",\\"segments\\":[{\\"origin\\":\\"MOW\\",\\"destination\\":\\"HKT\\",\\"date\\":\\"2021-04-03T08:09:00.000Z\\",\\"stops\\":[\\"IST\\",\\"BKK\\",\\"DXB\\"],\\"duration\\":855},{\\"origin\\":\\"HKT\\",\\"destination\\":\\"MOW\\",\\"date\\":\\"2021-04-22T23:47:00.000Z\\",\\"stops\\":[],\\"duration\\":1750}]}"
    price={43073}
    segments={
      Array [
        Object {
          "date": "2021-04-03T08:09:00.000Z",
          "destination": "HKT",
          "duration": 855,
          "origin": "MOW",
          "stops": Array [
            "IST",
            "BKK",
            "DXB",
          ],
        },
        Object {
          "date": "2021-04-22T23:47:00.000Z",
          "destination": "MOW",
          "duration": 1750,
          "origin": "HKT",
          "stops": Array [],
        },
      ]
    }
  />
  <Ticket
    carrier="EY"
    key="{\\"price\\":44751,\\"carrier\\":\\"EY\\",\\"segments\\":[{\\"origin\\":\\"MOW\\",\\"destination\\":\\"HKT\\",\\"date\\":\\"2021-04-03T01:37:00.000Z\\",\\"stops\\":[\\"IST\\",\\"HKG\\"],\\"duration\\":670},{\\"origin\\":\\"HKT\\",\\"destination\\":\\"MOW\\",\\"date\\":\\"2021-04-23T15:13:00.000Z\\",\\"stops\\":[\\"SIN\\",\\"DXB\\"],\\"duration\\":1289}]}"
    price={44751}
    segments={
      Array [
        Object {
          "date": "2021-04-03T01:37:00.000Z",
          "destination": "HKT",
          "duration": 670,
          "origin": "MOW",
          "stops": Array [
            "IST",
            "HKG",
          ],
        },
        Object {
          "date": "2021-04-23T15:13:00.000Z",
          "destination": "MOW",
          "duration": 1289,
          "origin": "HKT",
          "stops": Array [
            "SIN",
            "DXB",
          ],
        },
      ]
    }
  />
  <Ticket
    carrier="SU"
    key="{\\"price\\":79466,\\"carrier\\":\\"SU\\",\\"segments\\":[{\\"origin\\":\\"MOW\\",\\"destination\\":\\"HKT\\",\\"date\\":\\"2021-04-03T04:20:00.000Z\\",\\"stops\\":[\\"SIN\\",\\"SHA\\"],\\"duration\\":661},{\\"origin\\":\\"HKT\\",\\"destination\\":\\"MOW\\",\\"date\\":\\"2021-04-23T05:53:00.000Z\\",\\"stops\\":[\\"AUH\\"],\\"duration\\":1600}]}"
    price={79466}
    segments={
      Array [
        Object {
          "date": "2021-04-03T04:20:00.000Z",
          "destination": "HKT",
          "duration": 661,
          "origin": "MOW",
          "stops": Array [
            "SIN",
            "SHA",
          ],
        },
        Object {
          "date": "2021-04-23T05:53:00.000Z",
          "destination": "MOW",
          "duration": 1600,
          "origin": "HKT",
          "stops": Array [
            "AUH",
          ],
        },
      ]
    }
  />
</div>
`;
